sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/stats/SyncStatsRepository.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2023 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.sync.impl.stats
+
+import com.duckduckgo.sync.impl.engine.SyncStateRepository
+import com.duckduckgo.sync.store.model.SyncAttempt
+import com.duckduckgo.sync.store.model.SyncAttemptState.SUCCESS
+import java.math.RoundingMode
+import javax.inject.Inject
+
+/**
+ * The Repository that fetches all stats related to [SyncAttempt]
+ */
+interface SyncStatsRepository {
+
+    /**
+     * Returns yesterday's [DailyStats]
+     * @return [DailyStats]
+     */
+    fun getDailyStats(): DailyStats
+}
+
+data class DailyStats(
+    val attempts: Int,
+    val successRate: Double,
+)
+
+class RealSyncStatsRepository @Inject constructor(private val syncStateRepository: SyncStateRepository) : SyncStatsRepository {
+    override fun getDailyStats(): DailyStats {
```

Comment:
To confirm I understand this pixel:

I assume this pixel will report metrics about sync for each day. So the goal here is to know the attempts / rate / other data for a specific day.

I see we rely on work manager to trigger the pixel every 24h. I believe that will not give us what we want. Imagine workmanager triggers the worker every 24h at 9am. Then filtering by today will give us the attempts so far until that moment, but next day we will lose anything that happened the previous day between 9am until EoD. Is my assumption correct?

Reply:
I had this right in the API javadoc but not in the code. What we really want to send is yesterday’s statistics, not today. Your assumption is correct, we can’t rely on WorkManager to give us today’s data, but we can use WorkManager to schedule every 24 hours, and then get the data from the day before. That way we know that all data from the previous day has been properly stored.